### ISSUE TITLE
Filter expressions: confusing 'exists' semantics

### ISSUE THREAD

**Issue by bram209** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#180]
Considering that, according to the grammar, `exist-expr` is a `boolean-exrp`:

```
exist-expr        = [neg-op S] singular-path 
```

And the following note:
> A member or element value by itself in a Boolean context is interpreted as false only if it does not exist. Otherwise it is interpreted as true. To be more specific about the actual value, explicit comparisons are necessary. This existence test -- as an exception to the general rule -- also works with structured values.

So this would mean that a singular path that points to a boolean `false` value would evaluate to `true`.

```json
[
   { "id": 1, "isActive": true },
   { "id": 2, "isActive": false },
   { "id": 3, "isActive": true },
]
```
People might try:
`$[?(@.isActive)]` to get the active elements but will be surprised to see all elements returned.

I think this kind of query is quite common and for people with a developer background it would feel unnatural to type `boolean expression == true`

On the other side,  the need for queries that check *implicitly* for existence feels rarer to me (from my experience mostly well-typed tagged unions are used to structure JSON), which makes me wonder if `exists` should be a function instead (once functions get added to the spec).

This would make the `exists` check more explicit and therefore less confusing, in my opinion.

### Proposed solution
Personally, I would like to:
- have an explicit way of checking existence.
- evaluate singular path to a boolean value (which could be seen as a tautological statement) in the above example

When a singular path is given as a boolean expression but does not point to a boolean value, it gets a bit tricky:
- it could throw an error at the evaluation
- evaluate to false always
- introduce the concept of truthy & falsy values, which could serve similar use cases as `exists`, in this case, `true` would be `truthy` and `false` would be falsy and the above described problem would go away

I normally am more a statically typed programming language guy so not too keen on the concept of truthy & falsy values, but considering this is a query language for JSON, it might make more sense...

Thoughts?

**Comment by bram209** [comment]
another possibility (that does not rely on functions) is to introduce a new primitive value `undefined` so one would type:
`$[?(@.prop != undefined)]`

**Comment by glyn** [comment]
Unfortunately, there isn't much of a consensus among implementations: [comparison project example](https://cburgmer.github.io/json-path-comparison/results/filter_expression_with_value.html).

**Comment by cabo** [comment]
In a statically typed language with a Boolean-typed variable `x`,

```
if x == true {...}
```

definitely is a code smell, usually outing a complete noob.

But this isn't a statically typed language with a Boolean type.
(JSON is schema-free and thus dynamically typed, and there is nothing that even justifies the introduction of RFC 8259 to mention "booleans" (sic); the [actual definition](https://www.rfc-editor.org/rfc/rfc8259#section-3) only knows the three separate and independent values false, null, and true.)

**Comment by gregsdennis** [comment]
> JSON is schema-free and thus dynamically typed

This is just wrong.  _Javascript_ is dynamically typed because it considers objects with different property sets to be different types.  _JSON_ is statically typed: objects, arrays, strings, numbers, and the constants `true`, `false`, and `null`.  (Practically speaking, `true` and `false` are booleans.  Making this distinction is splitting hairs and not worth our time.)  For JSON, objects with different property sets, while they are not equivalent, are still the same type: `object`.

> introduce a new primitive value `undefined`...

We've had plenty of conversations stating that we don't want to introduce `undefined` as a value.  To reiterate the above, JSON isn't Javascript; there is no `undefined` in JSON.

> `if x == true {...}` definitely is a code smell

Agreed, but the argument is that for the expression `!x`, `x` _appears_ to be coerced to a boolean value, and then the `!` operator is negating that result.

We also have declared that expressions in filter selectors (`[?(<expr>)]`) must return boolean values.

This results in a syntactical ambiguity for `[?(!@.foo)]`:

- Is the user testing for the existence of a property?
- Is the user testing for the "truthiness" of the value at that property?

The _only_ way to know for sure is by forcing `!` followed by a local path (e.g. `!@.foo`) to mean an existence test, which means the _only_ way to check for "truthiness" is to compare with a `true` or `false` constant.

For what it's worth, my implementation, https://json-everything.net/json-path, returns all three items for the example above, acting as a true existence test.  This means that for my library, you MUST use `x == true` to only get the "truthy" ones.

---

To summarize, I see two options to resolve this ambiguity:

1. An expression in the form `!@.foo` is defined explicitly as an existence test, which forces path authors to use expressions like `x == true` for "truthiness" checks (would return three items from the above example).
2. Expressions strictly return booleans except in the special case of `!@.foo`, for which the expression is "truthy" or "falsy."  This forces some other form of existence test, e.g. a function form like `exists(@.foo)` (would return two items from the above example).

***NOTE** We've also discussed truthiness at length, and if I recall, we decided against it.  I use it here solely to illustrate my point.*

**Comment by glyn** [comment]
The current specification seems sufficient. Closing.
